Issue #2871698 by FeyP: Prevent exception upon saving a new entity, if tokens that rely on entity id are used

FeyP opened the pull request:
see [Drupal.org Issue #2871698](https://www.drupal.org/node/2871698)

stefanospetrakis commented on src/IcsFileManager.php:
```diff
@@ -119,6 +119,47 @@ public function updateIcalFile(ContentEntityBase $contentEntity, FieldDefinition
   }
 
   /**
+   * Create an empty ICS file for an entity.
```
Creates instead of create (nitpick)

stefanospetrakis commented on src/IcsFileManager.php:
```diff
@@ -119,6 +119,47 @@ public function updateIcalFile(ContentEntityBase $contentEntity, FieldDefinition
   /**
+   * Create an empty ICS file for an entity.
+   *
+   * We can't create the actual cotnents of an ICS file during pre-save, since
```
contents (typo)

stefanospetrakis commented on src/Plugin/Field/FieldType/CalendarDownloadType.php:
```diff
@@ -148,6 +159,7 @@ public function postSave($update) {
       // The current fielditem belongs to a fielditemlist,
       // that in turn belongs to a fieldable entity.
       $entity = $this->getParent()->getParent()->getValue();
+      $this->icsFileManager->updateIcalFile($entity, $this->getFieldDefinition(), $this->getValue());
```
If the entity is not new, then we call updateIcalFile twice, as it is already called once in preSave().
Could we make this conditional on whether the entity isNew()?
Or, find another way to avoid the double calls?

FeyP replied:
Actually, this code is executed only, if the `$update` parameter of `postSave()` is `FALSE`, i.e., if an entity is created. The check for that is in line 158, just outside the diff. Thus, `updateIcalFile()` is never called twice. For "old" entities, it will be executed during `preSave()` and for "new" entities, it will be executed during `postSave()`.

FeyP commented:
Thanks for the review. I updated the doc block as suggested and extended the existing text to use a token relying on entity id being present. This will make the test fail with the code currently in the 8.x-1.x branch, since the node wont be created. It will work fine with the code from this pull request. I didn't make any changes to the postSave method (see my comment above).

Let me know, if you want me to squash the commits into one later. For now, I added them on top of the original pull request to ease further review.

stefanospetrakis commented:
You're welcome and sorry for the delay.

I think we are good here, the only request I would have would be to repost this PR to the develop branch (and squash the commits while you are at it, if it's not a big problem for you).

Thanks!

FeyP commented:
Thanks! I squashed the commits into one and changed the target branch to `develop`.

stefanospetrakis commented:
Hey @FeyP , I just added codeship integration to the project, to allow running the tests from inside github. Would you mind rerolling against latest develop (in order to get the codeship configuration)?

BTW, this is a really cool combination, codeship for OS + https://github.com/ibrows/drupal_module_tester/

Thanks and cheers

stefanospetrakis commented:
Hey @FeyP, this fell asleep for while. Would you be able to reroll your commit against the latest dev? We are very very close. Thanks for your help and patience with this.

FeyP commented:
Hi @stefanospetrakis, sorry for not following up more quickly. I've been super busy with project work for the last two months and I have to admit, that I also forgot about this issue at a certain point, so thanks for the reminder ;). I'll guess, that this is a quick one, so I'll try to work this in as soon as possible, probably wont be today, but hopefully early next week. Thanks again!

FeyP commented:
Hi @stefanospetrakis, I just did the rebase. Sorry again for the delay.

stefanospetrakis commented:
Thanks @FeyP, it's merged now and the Codeship powered tests came back green. Thanks a ton for the work on this and the patience while trying to improve the module's code and tests. Will update the Drupal.org issue as well (and give credits).
